viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala

```diff
@@ -1873,33 +1873,426 @@ def setCalibrationPoints(startCp: Long, endCp: Long, projectLinks: Seq[ProjectLi
+      val leftTerminatedSections = toContinuousSectionsByStatus(terminatedLeft, RoadAddressChangeType.Termination)
+      val rightTerminatedSections = toContinuousSectionsByStatus(terminatedRight, RoadAddressChangeType.Termination)
+
+      val roadPartStartTerminatedLefts = leftTerminatedSections.find(section => section.exists(_.addrMRange.isRoadPartStart))
+      val roadPartStartTerminatedRights = rightTerminatedSections.find(section => section.exists(_.addrMRange.isRoadPartStart))
+
+      val processedLinks = if (roadPartStartTerminatedLefts.nonEmpty && roadPartStartTerminatedRights.nonEmpty) { // Check if road part start is terminated on both tracks
+        handleTwoTrackRoadPartStartTermination(roadPartStartTerminatedLefts.get, roadPartStartTerminatedRights.get, projectLinks)
+      } else
+        projectLinks
+
+      processedLinks
+    }
+
+    def roadPartEndTerminated(projectLinks: Seq[ProjectLink]): Seq[ProjectLink] = {
+      // TODO
+      projectLinks
+    }
+
+    def roadPartMiddleTerminated(projectLinks: Seq[ProjectLink]): Seq[ProjectLink] = {
+      val terminated = projectLinks.filter(_.status == RoadAddressChangeType.Termination)
+      val terminatedLeft = terminated.filter(_.track == Track.LeftSide)
+      val terminatedRight = terminated.filter(_.track == Track.RightSide)
+
+      // Group terminated links into continuous sections by status
+      val leftTerminatedSections = toContinuousSectionsByStatus(terminatedLeft, RoadAddressChangeType.Termination)
+      val rightTerminatedSections = toContinuousSectionsByStatus(terminatedRight, RoadAddressChangeType.Termination)
+
+      val processedLinks = if (leftTerminatedSections.nonEmpty && rightTerminatedSections.nonEmpty) { // Check if road part start is terminated on both tracks
+        handleTwoTrackMiddleTermination(leftTerminatedSections, rightTerminatedSections, projectLinks)
+      } else
+        projectLinks
+
+      processedLinks
+    }
+
+    def existsTerminationsOnTwoDiffTracks(projectLinks: Seq[ProjectLink]): Boolean = {
+      projectLinks.exists(pl => pl.track == Track.RightSide && pl.status == RoadAddressChangeType.Termination) &&
+      projectLinks.exists(pl => pl.track == Track.LeftSide && pl.status == RoadAddressChangeType.Termination)
+    }
+
+    // Check that there are terminated links on both tracks
+    val processedLinks: Seq[ProjectLink] = {
+      if (existsTerminationsOnTwoDiffTracks(projectLinks)) {
+        val orderedProjectLinks = projectLinks.sortBy(_.addrMRange.start)
+        
+        // List of termination cases
+        val terminationCases: List[Seq[ProjectLink] => Seq[ProjectLink]] = List(
+          roadPartStartTerminated,
+          roadPartEndTerminated,
+          roadPartMiddleTerminated
+        )
+
+        // Sequentially apply each case function
+        terminationCases.foldLeft(orderedProjectLinks)((linksAccumulator, caseFunction) => caseFunction(linksAccumulator))
+
+      } else {
```

Comment:
How about if there is a termination on one track, only?

That is either replaced by a new link (no problem), the non-terminated is now a combined (no problem), or no new link (error, but handled later)?

Reply:
Tämän funtion tehtävä on täsmäyttää lakkautetut 2-ajr osuudet, 
se ei huomioi sitä jos 2-osuudelta on lakkautettu vain toinen ajorata, eikä myöskään validoi onko lakkautukset valideja. (Validointi tapahtuu projektilinkkien etäisyyslukemien uudelleen laskennan jälkeen. )